lib/spack/spack/build_systems/cmake.py

```diff
@@ -246,3 +256,79 @@ def check(self):
 
     # Check that self.prefix is there after installation
     run_after('install')(PackageBase.sanity_check_prefix)
+
+    def write_spconfig(self, spack_env, spconfig_fname, dirty):
+        fname = os.path.join(spack_env.path, spconfig_fname) \
+            if spack_env is not None else spconfig_fname
+
+        with open(fname, 'w') as fout:
+            self._write_spconfig(fout, dirty)
+            fout.write('\nproc = subprocess.Popen(cmd, env=env)\n'
+                       'proc.wait()\n')
+
+        set_executable(fname)
+        return fname
+
+    def _write_spconfig(self, fout, dirty):
+        """Writes the spconfig.py file to a stream."""
```

Comment:
> There are two contexts for `dirty`, and the value of this variable will affect both of them:

I was not aware of the second meaning of `--dirty`.  Historically, Spack Setup has not enabled the use of the Spack wrappers (which didn't historically work outside of Spack).  In spite of this flaw, the feature was quite useful anyway.  But I agree, setting things up exactly the same as `spack build-env` would be better (and could also reduce duplicated functionality within Spack.)

> It also occurs to me that if you want `--dirty` in the sense that you pull all of the shell environment, `spconfig` itself could support an argument like that: if you invoke *it* with `--dirty` then it could update `os.environment` rather than overwrite it.

I think the right place for the dirty flag is in the `env.yaml` file.  My general logic behind this conclusion is: the reason people use `--dirty` is because, for whatever reason, a package doesn't build in a clean environment.  That's a property of the package and system Spack is running on.  If a package on a particular system requires `--dirty` once, then it will require it every time things are built / configured.  Users shouldn't have to remember that "package A requires dirty but package B does not."  Hence, it should be placed in `env.yaml` when needed.

> I don't necessarily want to hold up this PR on this point, but I think the following is worth discussing at least briefly:

I recommend we just merge as-is; and then follow on with  another PR that unifies `spack build-env` and `spack setup`.  I think it will take some thought and experimentation.  (Meaning: I put the `dirty` stuff in there because I needed it.  Trying to build my stuff will be a good test of any alternative we come up with.)